src/sentry/snuba/entity_subscription.py

```diff
@@ -235,40 +235,56 @@ def get_entity_extra_params(self) -> Mapping[str, Any]:
         return {"organization": self.org_id, "groupby": self.get_query_groupby()}
 
 
-def map_aggregate_to_entity_subscription(
+EntitySubscription = Union[
+    EventsEntitySubscription,
+    MetricsCountersEntitySubscription,
+    TransactionsEntitySubscription,
+    SessionsEntitySubscription,
+]
+ENTITY_KEY_TO_ENTITY_SUBSCRIPTION: Mapping[EntityKey, Type[EntitySubscription]] = {
+    EntityKey.Events: EventsEntitySubscription,
+    EntityKey.MetricsCounters: MetricsCountersEntitySubscription,
+    EntityKey.Sessions: SessionsEntitySubscription,
+    EntityKey.Transactions: TransactionsEntitySubscription,
+}
+
+
+def get_entity_subscription_for_dataset(
     dataset: QueryDatasets, aggregate: str, extra_fields: Optional[Mapping[str, Any]] = None
-) -> BaseEntitySubscription:
+) -> EntitySubscription:
     """
     Function that routes to the correct instance of `EntitySubscription` based on the dataset,
     additionally does validation on aggregate for datasets like the sessions dataset and the
     metrics datasets then returns the instance of `EntitySubscription`
     """
-    entity_subscription: BaseEntitySubscription
-    if dataset == QueryDatasets.SESSIONS:
-        match = re.match(CRASH_RATE_ALERT_AGGREGATE_RE, aggregate)
-        if not match:
-            raise UnsupportedQuerySubscription(
-                "Only crash free percentage queries are supported for subscriptions"
-                "over the sessions dataset"
-            )
-        entity_subscription = SessionsEntitySubscription(aggregate, extra_fields)
-    elif dataset == QueryDatasets.TRANSACTIONS:
-        entity_subscription = TransactionsEntitySubscription(aggregate, extra_fields)
-    elif dataset == QueryDatasets.METRICS:
+    return ENTITY_KEY_TO_ENTITY_SUBSCRIPTION[map_aggregate_to_entity_key(dataset, aggregate)](
+        aggregate, extra_fields
+    )
+
+
+def map_aggregate_to_entity_key(dataset: QueryDatasets, aggregate: str) -> Optional[EntityKey]:
+    entity_key = None
+    try:
+        entity_key = {
+            QueryDatasets.EVENTS: EntityKey.Events,
+            QueryDatasets.TRANSACTIONS: EntityKey.Transactions,
+        }[dataset]
+    # If we get to this, then dataset must be either the metrics or the sessions datasets
+    except KeyError:
```

Comment:
Using `KeyError` feels a bit weird here. Would we better defining the mapping outside and for all entities/datasets, then doing:
```
entity_key = entity_key_mapping[dataset]
if entity_key in [EntityKey.MetricsCounters, EntityKey.Sessions]:
    # Do extra error checking
return entity_key
```

Reply:
Yeah you are right!